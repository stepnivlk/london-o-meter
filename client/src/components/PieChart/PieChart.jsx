// @flow

import React from 'react';
import { arc } from 'd3';
import Smiley from '../Smiley';
import { pieConfig, labels } from '../../constants';

type Item = {
  label: string,
  value: number,
  index: number,
  valuePercent: number,
  color: string,
  isActive: boolean,
};

type Props = {
  data: Item[],
  dimensions: {
    width: number,
    height: number,
  },
  onPieHover: (number) => void,
  onPieHoverOut: () => void,
  arcs: Object[],
  highlightedIndex: number,
};

const createPieElement = (arcs, index, highlightedIndex) => {
  const radius = highlightedIndex === index ?
    pieConfig.RADIUS + 10 :
    pieConfig.RADIUS;

  const pieElement = arc()
    .outerRadius(radius)
    .padAngle(.05)
    .innerRadius(pieConfig.INNER_RADIUS);

  const arcData = arcs[index];

  return pieElement(arcData);
}

const winningItem = (data) => data.reduce((prev, current) =>
  (prev.value > current.value) ? prev : current, { label: '', value: 0 });

const isPositive = (data) =>
  winningItem(data).label === labels.POSITIVE;

const PieChart = ({
  data,
  dimensions: { width, height },
  onPieHover,
  onPieHoverOut,
  arcs,
  highlightedIndex
}: Props) => (
  <svg
    key="svg"
    width="100%"
    height={height}
    fontFamily="Helvetica"
  >
    <g transform={`translate(${width / 2},${height / 2})`}>
      {
        data.map((item) => (
          <g key={`${item.label}_index`}>
            <path
              d={createPieElement(arcs, item.index, highlightedIndex)}
              fill={item.color}
              onMouseEnter={() => onPieHover(item.index)}
              onMouseLeave={() => onPieHoverOut()}
            />
          </g>
        ))
      }
      <Smiley isPositive={isPositive(data)} />
    </g>
  </svg>
)

export default PieChart;
