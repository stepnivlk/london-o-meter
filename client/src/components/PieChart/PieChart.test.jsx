import React from 'react';
import renderer from 'react-test-renderer';
import PieChart from './PieChart';

const item = {
  label: 'positive',
  value: 10,
  valuePercent: 50,
  index: 0,
  color: '#fff',
  isActive: false
};

const arc = {
  data: item,
  index: 0,
  value: 33,
  startAngle: 0,
  endAngle: 3.3991002481463335,
  padAngle: 0
};

it('matches a snapshot', () => {
  const component = renderer.create(
    <PieChart
      data={[item]}
      dimensions={{ width: 100, height: 100 }}
      onPieHover={() => null}
      onPieHoverOut={() => null}
      arcs={[arc]}
      highlightedIndex={0}
    />
  );

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
