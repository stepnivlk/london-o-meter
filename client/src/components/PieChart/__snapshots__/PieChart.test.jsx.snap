// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches a snapshot 1`] = `
<svg
  fontFamily="Helvetica"
  height={100}
  width="100%"
>
  <g
    transform="translate(50,50)"
  >
    <g>
      <path
        d="M5.617466068175438,-209.92485339990807A210,210,0,1,1,-48.02955215083016,204.4337597369639L-14.8911547298741,78.60186709494214A80,80,0,1,0,5.617466068175428,-79.80253175791415Z"
        fill="#fff"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      />
    </g>
    <g
      transform="translate(-50,-50)"
    >
      <circle
        cx={25}
        cy={30}
        fill="black"
        id="c"
        r={5}
      />
      <circle
        cx={75}
        cy={30}
        fill="black"
        id="c"
        r={5}
      />
      <g
        transform="translate(15,65)"
      >
        <path
          d="M0,0 A40,40 10 0,0 65,0"
          fill="none"
          stroke="black"
          strokeWidth={5}
        />
      </g>
    </g>
  </g>
</svg>
`;
